internal/route/request_executor.go

```diff
@@ -50,23 +56,122 @@ func (r *RequestExecutor) routeToConfig(
 		httpReq.Header.Set("Authorization", "Basic "+encodedCredentials)
 	}
 
+	var (
+		respBody jsonrpc.ResponseBody
+		resp     *http.Response
+	)
+
+	singleRequestBody, isSingleRequestBody := requestBody.(*jsonrpc.SingleRequestBody)
+
+	if r.cache != nil && r.cache.ShouldCacheMethod(requestBody.GetMethod()) && isSingleRequestBody {
+		// In case of unknown caching errors, the httpReq might get used twice.
+		// We must clone the httpReq otherwise the body will already be closed on the second request.
+		respBody, resp, err = r.retrieveOrCacheRequest(cloneRequest(httpReq), *singleRequestBody, configToRoute)
+		if err != nil {
+			originError, _ := err.(*OriginError)
+			// An OriginError indicates a cache miss and request failure to origin.
+			// We want this error to bubble up.
+			// Any other errors should be unhandled caching errors.
+			// In these cases we default back to the "non-caching" behavior.
+			if originError != nil {
+				r.logger.Warn("caching error making request to origin", zap.Error(err), zap.Any("request", requestBody))
+				return nil, nil, originError
+			}
+		} else {
+			return respBody, resp, nil
+		}
+	}
+
+	respBody, resp, err = r.getResponseBody(httpReq, requestBody, configToRoute)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return respBody, resp, nil
+}
+
+func (r *RequestExecutor) retrieveOrCacheRequest(httpReq *http.Request, requestBody jsonrpc.SingleRequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *http.Response, error) {
+	var (
+		respBody jsonrpc.ResponseBody
+		resp     *http.Response
+	)
+
+	originFunc := func() (*jsonrpc.SingleResponseBody, error) {
+		var err error
+
+		respBody, resp, err = r.getResponseBody(httpReq, &requestBody, configToRoute) //nolint:bodyclose // linter bug
+		if err != nil {
+			return nil, err
+		}
+
+		singleRespBody, ok := respBody.(*jsonrpc.SingleResponseBody)
+		if !ok {
+			return nil, errors.New("batched responses do not support caching")
+		}
+
+		return singleRespBody, err
+	}
+	result, err := r.cache.HandleRequest(requestBody, originFunc)
+
+	// A cache hit or the request to origin failed.
+	if resp == nil && respBody == nil {
+		// Request to origin failed, we should fail out here.
+		if err != nil {
```

Comment:
On similar note, could we put `respBody` and `resp` defined on L95-96 as local vars in the `originFunc`? And then after getting the response we can initialize new variables for `respBody` and `resp` depending on the `result`. 

I was having a difficult time with tracking how these variables are getting used in & out of the closure .

Reply:
Yeah it was extremely difficult trying to fit the caching logic into this handler because the return types (ResponseBody and Response) are different than what is being cached (RawMessage).

Due to this, I can't find a way to set respBody and resp without using closures.  Whatever is returned by the cache `Do` is being marshaled and set into `Item.Value`, so I don't have many options here.

The main thing making it difficult to represent a cache hit is that I have four different scenarios:
1. Cache hit (respBody and resp need to get "created")
2. Cache miss and request to origin succeeds.  Cache and return the response from the origin.
3. Cache miss and request to origin fails.  Don't cache and return the error back to the user.
4. Cache miss and the request to origin succeeds BUT the Error field is set in the JSON RPC response.  Don't cache BUT return the response back to the user.

Making sure I handle all of these cases makes the logic difficult to write.  For example, I can't just use `if result != nil` to represent a cache hit because `result = nil` has three logic branches to handle.

If I move the `if err != nil ` check above the `if resp == nil && respBody == nil`, I would have the same level of nested conditionals because I need to check resp and respBody.

Reply:
Another difficulty is that resp and respBody are separate objects.  In order to determine the cache value, I have to both make the response and also decode it into the SingleResponseBody to extract out the `Result` field.  However, both resp and respBody also need to be returned!  So I have to use closures to set the values from inside the caching logic.

I think the redis cache library is also not making this any simpler, but I think we really need the `Once()` behavior to ensure a good hit rate.  `Once()` guarantees that identical requests only have a single in-flight request to the origin, which is very valuable in our use case because the indexers are very closely synced up during block ingestion across chains.

Reply:
Oh gotcha, thanks for the detailed explanation. 

I think #4 makes things tricky.

Does #2 behave the same as #1? Like if we miss cache and execute the request, will the `result`s be the same across case #1 and #2?

Reply:
In the future maybe we could consolidate resp and respBody into a single struct. May be easier to work with.

Reply:
#1 and #2 are slightly different.

In #1, `result` will be set with just the JSONRPC `Result` field contents.  We have to generate a fake resp and respBody to match the request.

In #2, we can directly return the resp and respBody that came from the request to the origin.

Technically though, I think we could reconstruct the resp and respBody in both cases?

Yeah I tried comboing them into a struct but I didn't find it actually helping to make the logic simpler.

Reply:
> nit: is it possible to move this above of the if resp == nil && respBody == nil ? I find reading multiple conditions under the if... block kind of hard to read.

I realized that I can actually do this as long as I classify a JSON RPC error as its own custom error.  Then I can return an error from `HandleRequest` only on "real" errors.  A little cleaner now!

Reply:
Sweet